tenseal/tensealcontext.h

```diff
@@ -62,11 +62,63 @@ class TenSEALContext {
         return shared_ptr<TenSEALContext>(new TenSEALContext(filename));
     }
 
-    shared_ptr<KeyGenerator> key_gen() { return this->_keygen; }
-    PublicKey public_key() { return this->_keygen->public_key(); }
-    SecretKey secret_key() { return this->_keygen->secret_key(); }
-    RelinKeys relin_keys() { return this->_keygen->relin_keys(); }
-    GaloisKeys galois_keys() { return this->_keygen->galois_keys(); }
+    PublicKey public_key() { return PublicKey(*this->_public_key); }
+    SecretKey secret_key() {
+        if (this->_secret_key == nullptr) {
+            throw invalid_argument(
+                "the current context is public, it doesn't hold a Secret key");
+        }
+
+        return SecretKey(*this->_secret_key);
+    }
+
+    RelinKeys relin_keys() {
+        if (this->_relin_keys == nullptr) {
+            throw invalid_argument(
+                "the current context doesn't hold a Relinearization keys");
+        }
+
+        return RelinKeys(*this->_relin_keys);
+    }
+
+    GaloisKeys galois_keys() {
+        if (this->_galois_keys == nullptr) {
+            throw invalid_argument(
+                "the current context doesn't hold a Galois keys");
+        }
+
+        return GaloisKeys(*this->_galois_keys);
+    }
+
+    /*
+    Generate Galois and Relinearization keys if needed, then destroy the
+    _secret_key and set it to nullptr
+    */
+    void make_context_public(bool generate_galois_keys,
+                             bool generate_relin_keys) {
+        // create KeyGenerator object only if needed
+        if (generate_galois_keys || generate_relin_keys) {
+            KeyGenerator keygen = KeyGenerator(
+                this->_context, *this->_secret_key, *this->_public_key);
+
+            // generate Galois Keys
+            if (generate_galois_keys) {
+                this->_galois_keys = shared_ptr<GaloisKeys>(
+                    new GaloisKeys(keygen.galois_keys()));
+            }
+
+            // generate Relinearization Keys
+            if (generate_relin_keys) {
+                this->_relin_keys =
+                    shared_ptr<RelinKeys>(new RelinKeys(keygen.relin_keys()));
+            }
+        }
+
+        // destory and set _secret_key and decryptor to null
+        this->_secret_key = nullptr;
+        this->decryptor = nullptr;
+        this->_is_public = true;
+    }
 
     bool is_public() { return this->_is_public; }
     bool is_private() { return !this->_is_public; }
```

Comment:
I think we can now get rid of the _is_public attribute and base these functions on either _secret_key in nullptr